enhanced-stack-trace-queries.md

```diff
@@ -0,0 +1,84 @@
+# Enhanced stack trace queries
+
+## Backround
+
+The SDKs have a built-in `__stack_trace` query that can be used display where the workflow is blocked at the time of the
+query (the last processed workflow task).
+
+The response to a `__stack_trace` query is a single string representing a list of stack individual stack traces
+separated by double newlines.
+
+Example from the TS SDK:
+
+```
+    at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)
+    at someWorkflowFunction (/path/to/workflows/my-workflow.ts:60:20)
+
+    at someInternalFunction2 (/path/to/workflows/my-workflow.ts:100:30)
+    at someWorkflowFunction2 (/path/to/workflows/my-workflow.ts:123:5)
+```
+
+While this is valuable for developers to be able to understand how their code correlates to event histories, it is quite
+limited since it requires looking up the actual code to get the full picture.
+
+## Proposal
+
+SDKs can optionally implement an `__enhanced_stack_trace` query (name TBD) where instead of returning the stack traces
+as a single string, they return a pre-defined structure, that structure will also include file sources.
+
+The web UI can use this response to show the workflow code and highlight locations where it is blocked.
+
+### Proposed structure
+
+```ts
+/**
+ * A pointer to a location in a file
+ */
+export interface FileLocation {
+  filename: string;
+  line: number;
+  column: number;
+}
+
+/**
+ * Used as the result for the enhanced stack trace query
+ */
+export interface EnhancedStackTrace {
+  /**
+   * Mapping of file name to file contents
+   */
+  sources: Record<string, string>;
+  locations: FileLocation[];
```

Comment:
Shouldn't it be `FileLocation[][]`, ie. the enhanced stack trace for each blocked promise? Or do you intend to only return the top most location for each blocked promise?

Reply:
I was intending to return the top-most location for each blocked promise but that's a good point it'd be valuable to know the stack, just not sure how to visualize that, @stevekinney

Reply:
cc @canvanooo

Reply:
Agreed here. It's not a stack trace without... the stack :smile: 

I think you might want to ship slightly more data so there's at least something valuable even if sources aren't provided.

Ex:

```ts
export interface StackFrame {
    location: FileLocation
    // This can be the "someInternalFunction" part of "at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)"
    context: String 
    // We can add stuff here in a backwards compat manner
}

export interface StackTrace {
    stack: StackFrame[]
}
```

Then this line becomes
`stacks: StackTrace[]`

Reply:
Not sure what would go in context but it's a good idea to make `StackFrame` an object.

I'll definitely update the proposal to something like this.

We could a "step out" button in the UI to jump to the caller's location.

We don't have to support this immediately in the UI but it should be trivial to add and having the full trace definitely adds value.

Thanks both for pointing this out.

Reply:
Another nice thing to add would be some details would allow linking the promise to a corresponding workflow history event, when appropriate…

Reply:
Yes, that is something I want to add as well but those aren't directly connected to workflow stack traces which show where the user code is blocked.
I did want to include for each event, the stack trace that lead to creating that event but the plan was to do it in the "advanced" proposal, the one that requires worker queries so it can be done in all SDKs.

Reply:
It should be easy enough to do in TS since we already capture all of these stack traces but we don't in other SDKs at the moment.
I'm debating whether I should add it in this proposal or not.

Reply:
I'll keep this thread alive for reference.

Reply:
> Not sure what would go in context 

Like in my comment - you often get something like the function name when you dump the stack. That can be useful to see, especially if full source isn't available.

Reply:
Yeah, I figured that out after commenting, for some reason I missed the intent there. It's now a part of FileLocation.